influence_benchmark/stats/retroactive_evals.py

```diff
@@ -0,0 +1,172 @@
+                    for j, device in enumerate(self.devices):
+                        start = i + j * self.batch_size
+
+                        end = min(start + self.batch_size, total_transcripts)
+
+                        # Break if we've processed all transcripts
+                        if start >= total_transcripts:
+                            break
+
+                        # Append a tuple containing the batch, start index, and device
+                        device_batches.append((all_transcripts[start:end], start, device))
+
+                    if not device_batches:
+                        break
+
+                    # Process batches in parallel using starmap
+                    batch_results = pool.starmap(self.evaluate_batch, device_batches)
+
+                    # Extend results with processed batches
+                    results.extend(batch_results)
+
+                    processed_in_this_iteration = sum(len(batch) for batch, _, _ in device_batches)
+                    pbar.update(processed_in_this_iteration)
+
+        # Flatten the results from all batches
+        flat_results = list(itertools.chain(*results))
+        # Sort results by their original index to maintain order
+        sorted_results = sorted(flat_results, key=lambda x: x[0])
+
+        # Create a copy of the original DataFrame to store evaluation results
+        traj_df_evals = self.traj_df.copy()
+
+        # Add new columns for each metric
+        for metric in self.metrics:
+            traj_df_evals[metric] = [result[1][metric] for result in sorted_results]
+
+        # Store the evaluation results
+        self.traj_df_evals = traj_df_evals
+        return traj_df_evals
+
+    def evaluate_batch(self, batch, start_index, device):
+        # Create a new backend for this process with the specified device
+        process_backend = HFBackend(
+            model_name=self.backend_config["model_name"],
+            model_id=self.backend_config["model_id"],
+            lora_path=self.backend_config["lora_path"],
+            device=device,
+        )
+
+        # Evaluate each transcript in the batch
+        return [
+            (start_index + i, self.evaluate_transcript(transcript, process_backend))
+            for i, transcript in enumerate(batch)
+        ]
+
+    def evaluate_transcript(self, transcript, backend):
+        # Evaluate the transcript for each metric using the provided backend
+        return {metric: self.get_eval_from_backend(transcript, metric, backend) for metric in self.metrics}
+
+    def get_eval_from_backend(self, transcript, metric, backend):
+        state = RetroactiveIterationState(history=transcript, variables=self.state_variables)
+        messages = self.assessor_models[metric].prepare_messages(state)
+        valid_tokens = self.config[metric]["valid_tokens"]
+        responses = backend.get_next_token_probs_normalized_vec([messages], valid_tokens_n=[valid_tokens])
```

Comment:
might be cleaner if you go via vectorized assesor model, I think that's where you put your normalization code as well. Probably not that important

Reply:
I see, yes. For now, I think it is OK because it will not change the functionality too much, but would be happy to make this change later if we want. It may interact in intricate ways with the mp/async code I set up right now so that's why I feel like it's not worth a ton of time. 

For the normalization, I moved those functions to a utils file so I can call them here too